xdc-plugin/xdc.cc

```diff
@@ -0,0 +1,268 @@
+	}
+
+	void execute(std::vector<std::string> args, RTLIL::Design* design) YS_OVERRIDE {
+		if (design->top_module() == nullptr) {
+			log_cmd_error("No top module detected\n");
+		}
+
+		std::string option(args[1]);
+		if (set_property_options_map.count(option) == 0) {
+			log_warning("set_property: %s option is currently not supported\n", option.c_str());
+			return;
+		}
+
+		switch (set_property_options_map.at(option)) {
+			case SetPropertyOptions::INTERNAL_VREF:
+				process_vref(std::vector<std::string>(args.begin() + 2, args.end()), design);
+				break;
+			default:
+				assert(false);
+		}
+	}
+
+	void process_vref(std::vector<std::string> args, RTLIL::Design* design) {
+		if (args.size() < 2) {
+			log_error("set_property INTERNAL_VREF: Incorrect number of arguments.\n");
+		}
+		int iobank = std::atoi(args[1].c_str());
+		auto bank_tiles = get_bank_tiles();
+		if (bank_tiles.count(iobank) == 0) {
+			log_cmd_error("set_property INTERNAL_VREF: Invalid IO bank.\n");
+		}
+
+		int internal_vref = 1000 * std::atof(args[0].c_str());
+		if (internal_vref != 600 &&
+				internal_vref != 675 &&
+				internal_vref != 750 &&
+				internal_vref != 900) {
+			log("set_property INTERNAL_VREF: Incorrect INTERNAL_VREF value\n");
+			return;
+		}
+
+		// Create a new BANK module if it hasn't been created so far
+		RTLIL::Module* top_module = design->top_module();
+		if (!design->has(ID(BANK))) {
+			std::string fasm_extra_modules_dir(proc_share_dirname() + "/plugins/fasm_extra_modules");
+			Pass::call(design, "read_verilog " + fasm_extra_modules_dir + "/BANK.v");
+		}
+
+		// Set parameters on a new bank instance or update an existing one
+		char bank_cell_name[16];
+		snprintf(bank_cell_name, 16, "\\bank_cell_%d", iobank);
+		RTLIL::Cell* bank_cell = top_module->cell(RTLIL::IdString(bank_cell_name));
+		if (!bank_cell) {
+			bank_cell = top_module->addCell(RTLIL::IdString(bank_cell_name), ID(BANK));
+		}
+		bank_cell->setParam(ID(FASM_EXTRA), RTLIL::Const("INTERNAL_VREF"));
+		bank_cell->setParam(ID(NUMBER), RTLIL::Const(iobank));
+		bank_cell->setParam(ID(INTERNAL_VREF), RTLIL::Const(internal_vref));
+	}
+
+	std::function<const BankTilesMap&()> get_bank_tiles;
+};
+
+struct ReadXdc : public Frontend {
+	ReadXdc()
+	       	: Frontend("xdc", "Read XDC file")
+		, GetIOBanks(std::bind(&ReadXdc::get_bank_tiles, this))
+		, SetProperty(std::bind(&ReadXdc::get_bank_tiles, this)) {}
+
+	void help() YS_OVERRIDE {
+		//   |---v---|---v---|---v---|---v---|---v---|---v---|---v---|---v---|---v---|---v---|
+		log("\n");
+		log("    read_xdc -part_json <part_json_filename> <filename>\n");
+		log("\n");
+		log("Read XDC file.\n");
+		log("\n");
+	}
+
+	void execute(std::istream *&f, std::string filename, std::vector<std::string> args, RTLIL::Design*) YS_OVERRIDE {
+                if (args.size() < 2) {
+                        log_cmd_error("Missing script file.\n");
+		}
+                Tcl_Interp *interp = yosys_get_tcl_interp();
+		size_t argidx = 1;
```

Comment:
You should clear bank_tiles if -part_json is not supplied to prevent stale data from passing from call to call.

Reply:
Done